Committed changes for forgot password request feature (#1 of 2).

harshadajog opened the pull request:
### :scroll: Description

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
New Feature: https://trello.com/c/XC1mShgs/101-eng-forgot-password

- Describe what changes are being made
1. Added new route to forgot-password
2. Added ForgotPasswordRequest component
3. Added necessary inline validations for email address (missing, invalid, email not registered in system)
4. Added BE logic to generate password token for requesting user & sending email.
5. Added mutation to update password_reset_token in DB

- Describe why these changes are being made
This is for the new feature: https://trello.com/c/XC1mShgs/101-eng-forgot-password

- List the use cases and edge cases relevant to this PR
1. Missing or Invalid email check
2. Case where the user account does not exist. Show relevant error
3. Happy Case - User requests forgot password. Receives email with link with token.

- Describe how errors will be handled. How will we know if this code breaks in production
Cases are handled in Frontend before calling BE. To minimize breaking.
Test cases are covered in above step.

- Describe any external libraries/dependencies added or removed in this PR
None

- Describe any security risks are there regarding this change
<!--- Add your answer here -->

- Describe how you tested these changes
Manual testing of each flow.
--------------------------
### :clipboard: Mandatory Checklist

- [x] Example of a checked item (please remove when creating your Pull Request) 

- [ x] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [ x] Have you discussed these changes with the project leader(s)?
- [ x] Do all variable and function names communicate what they do?
- [x ] Were all the changes commented and / or documented?
- [x ] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [x ] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?
- [x ] *If you changed the user interface, did you add before and after screenshots to below?*

--------------------------
### :framed_picture: Screenshots and Screen Recordings

#### Before
<!--- Add before screenshots here -->
New

#### After
<!--- Add after screenshots here -->

![image](https://user-images.githubusercontent.com/3578183/141020018-e2c74378-6e80-4286-9fae-3c5b5a4889d5.png)

![image](https://user-images.githubusercontent.com/3578183/141020088-3e03659a-157f-4d2f-a4e2-8e50249613fd.png)

--------------------------
### :blue_book: Glossary
- PR = Pull Request

jdunbrack commented on frontend/src/components/ForgotPassword/ForgotPasswordRequest.styles.scss:
```diff
@@ -0,0 +1,153 @@
+            }
+        }
+    }   
+}
```
nit: sass file names so far have been *.scss, as opposed to *.styles.scss

harshadajog replied:
Done

jdunbrack commented on frontend/src/components/ForgotPassword/ForgotPasswordRequest.jsx:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+      const eraseFieldError = (element) => {
```
Does the `element` arg here do anything? I see the focus handler below calls it with `e.target.name` but it's not clear where that goes.

harshadajog replied:
No it doesn't. Updating as per above comments.

jdunbrack commented on frontend/src/components/ForgotPassword/ForgotPasswordRequest.jsx:
```diff
@@ -0,0 +1,110 @@
+                                placeholder={t('forgotPassword.body.input.email')}
+                                value={email}
+                                onChange={(e) => setEmail(e.target.value)}
+                                onFocus={(e) => eraseFieldError(e.target.name)}
```
Focus handler mentioned above.

harshadajog replied:
Updating as per above comments

jdunbrack commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -116,7 +116,8 @@ function LoginHandler() {
               onChange={(e) => setPassword(e.target.value)}
             />
-            <a className="passAnchor desktopView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a>
+            {/* <a className="passAnchor desktopView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a> */}
```
Extra comment

jdunbrack commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -125,10 +126,11 @@ function LoginHandler() {
             </button>
-            <a className="passAnchor mobileView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a>
+            <NavLink className="passAnchor mobileView" to="/forgot-password">{t('login.body.textSignIn.forgotPass')}</NavLink>
+            {/* <a className="passAnchor mobileView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a> */}
```
Extra comment

harshadajog replied:
Removed

jdunbrack commented on frontend/src/components/LoginHandler/LoginHandler.jsx:
```diff
@@ -116,7 +116,8 @@ function LoginHandler() {
             />
-            <a className="passAnchor desktopView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a>
+            {/* <a className="passAnchor desktopView" href="#passAnchor">{t('login.body.textSignIn.forgotPass')}</a> */}
+            <NavLink className="passAnchor desktopView" to="/forgot-password">{t('login.body.textSignIn.forgotPass')}</NavLink>
```
In Figma, Desktop uses "Forget Password?" but mobile uses "Forgot Password?"

I assume this is a design error and they should be the same, as you have it coded here?

harshadajog replied:
I believe so. Also, the feature will have new designs as per latest update.

jdunbrack commented on frontend/src/components/UserAccountCreate/AccountCreate.jsx:
```diff
@@ -15,7 +15,7 @@ import {
 
   // TODO: Add Site key.
-  const RECAPTCHA_SITE_KEY = '';
+  const RECAPTCHA_SITE_KEY = process.env.REACT_APP_RECAPTCHA_SITE_KEY;
```
Similar to above, is this something we need to add to our .env files? I don't see this value in any of them so far.

harshadajog replied:
Yes, we will need to add it in .env. Afaik, the .env we have is a brief initial file. We will keep adding new things there (vs source code) as we build features. Our eng team will have to co-ordinate (maybe setup a process for keeping it together)

jdunbrack replied:
Sounds good, thank you.

jdunbrack reviewed (commented):
Not familiar with how strict the group is with code reviews here, so I made more comments than necessary with things I spotted. Feel free to let me know if any of these are not relevant!

harshadajog commented on backend/graphql_api/lambda/graphql/resolvers/mutation.js:
```diff
@@ -5,6 +5,9 @@ const getSubscriptionController = require('../../controllers/subscriptionsContro
 const getsesClient = require('../../ses/sesClient');
+const jwt = require('jsonwebtoken');
+const JWTSecret = process.env.JWT_SECRET; 
+const CLIENT_URL = process.env.CLIENT_URL; 
```
CLIENT_URL is not currently in .env files.

The ses client needs to send  the reset password link of the sort:
<domain_name>/reset-password?token=${token}`;
User then clicks the link and goes ahead.

We don't have a domain yet, so instead of adding something like "http://localhost:3001/reset-password?token=" or an incomplete link ...I thought this maybe a better option.

In case you have an alternate ways, please do let me know.

jdunbrack replied:
No that sounds great, I just didn't know where the value originated from, if it was passed to the app via the server, was written in a .env file somewhere, etc.

I like this as a solution and will remember it for myself later, thank you.

harshadajog commented on backend/graphql_api/lambda/graphql/resolvers/mutation.js:
```diff
@@ -307,5 +310,41 @@ module.exports = (logger) => {
+    dbClient, { emailAddress }, context,
+  ) => {
+    logger.info(`Backend: Inside forgotPassword mutation resolver ${emailAddress}`);
+    try {
+    const lowerCaseEMailAddress = emailAddress.toLowerCase().trim();
+    // Verify if user exists in the system.
+    const dbResponse = await dbClient.getAccountByEmail(lowerCaseEMailAddress);
+    if (dbResponse.rows.length == 0) {
+        logger.error('This user is not registered in our system.');
+        throw new Error('This user is not registered in our system.');   
+    } 
+    const password = dbResponse.rows[0].password;
+    const id = dbResponse.rows[0].id;
+
+    // Step 1: Create a jwt token with expiry time 
+    const token = jwt.sign(
+        {_id:id}, JWTSecret, { expiresIn: '20m' });
+
+    /* Step 2: Call DB Client.
+        Update account table for that accountID, set password_reset_token to token       
+    */
+    const dbUpdateResponse = await dbClient.updatePasswordResetTokenForAccount(id, token);
+
+    // Step 3: Generate password reset link.
+    const link = `${CLIENT_URL}/reset-password?token=${token}`;
+    logger.info(`Backend: Reset Password Link: ${link}`);
+      
+    // Step 6: Call Send email
+    sesClient.sendForgotPasswordEmail(emailAddress, link);
+      return link;
+    } catch (e) {
+    logger.error(`forgotPassword mutation resolver error -  ${e}`);
+    throw new Error(e);
+    }  
```
Updating. Thanks. Will push the changes soon.

harshadajog commented:
Hi @jdunbrack. I have updated as per review comments & also added next part of the feature.
Submitted it as part of #253. This closes the https://trello.com/c/XC1mShgs/101-eng-forgot-password

I am closing this request.
